core/modules/path/path.module

```diff
@@ -110,9 +154,282 @@ function path_menu() {
+      'page callback' => 'backdrop_goto',
+      'page arguments' => array('admin/config/search/redirect/404'),
+      'access arguments' => array('administer redirects'),
+      'type' => MENU_LOCAL_ACTION,
+    );
+  }
+
+  // Devel generate integration.
+  if (module_exists('devel_generate')) {
+    $items['admin/config/development/generate/redirects'] = array(
+      'title' => 'Generate redirects',
+      'description' => 'Generate a given number of redirects. Optionally delete current redirects.',
+      'page callback' => 'backdrop_get_form',
+      'page arguments' => array('redirect_generate_form'),
+      'access arguments' => array('administer redirects'),
+      'file' => 'redirect.generate.inc',
+    );
+    $items['admin/config/search/redirect/generate'] = $items['admin/config/development/generate/redirects'];
+    $items['admin/config/search/redirect/generate']['type'] = MENU_LOCAL_ACTION;
+  }
+
   return $items;
 }
 
+function redirect_set_current_redirect($redirect) {
+  $static = &backdrop_static(__FUNCTION__);
+  $static = $redirect;
+}
+
+function redirect_get_current_redirect() {
+  $redirect = backdrop_static('redirect_set_current_redirect', NULL);
+
+  // If a redirect has not been set with redirect_set_current_redirect(), then
+  // attempt to find a redirect matching the current path, query string, and
+  // language code.
+  if (!isset($redirect)) {
+    $redirect = redirect_load_by_source(current_path(), $GLOBALS['language']->langcode, backdrop_get_query_parameters());
+  }
+
+  // @todo Add an alter hook here?
+  return $redirect;
+}
+
+/**
+ * Implements hook_url_inbound_alter().
+ */
+function path_url_inbound_alter(&$path, $original_path, $path_language) {
+  // If the current path global does not exist, then backdrop_get_path_alias()
+  // will fail. This condition only happens when $path is the front page.
+  // @todo Remove when http://drupal.org/node/1329914 is fixed in core.
+  if (empty($_GET['q'])) {
+    $_GET['q'] = config_get('system.core', 'site_frontpage');
+    return;
+  }
+
+  // If the inbound path has been changed, then attempt to find a redirect
+  // matching the original path and save it for processing later in
+  // redirect_init(). For example, if the Sub-pathauto module changes the path
+  // 'foo/redirect' to 'node/1/redirect', and there is a redirect enabled for
+  // the path 'foo/redirect', then redirect_init() would normally fail since it
+  // would not find a match.
+  if ($path != $original_path && $original_path == current_path()) {
+    $current_langcode = !empty($path_language) ? $path_language : $GLOBALS['language']->langcode;
+    $current_query = backdrop_get_query_parameters();
+    if ($redirect = redirect_load_by_source($original_path, $current_langcode, $current_query)) {
+      redirect_set_current_redirect($redirect);
+    }
+  }
+
+  // Redirect to canonical URLs.
+  // Commented out per https://www.drupal.org/node/2048137.
```

Comment:
Interesting. We should either uncomment this and try to fix it per that issue, it seems to be fixed by https://www.drupal.org/node/2400539.

Or we should remove this functionality entirely and solve it in a follow-up issue.